File: apps/burner-mac/Sources/FlagshipBurnerCore/BurnerPairing.swift
import Foundation
import CryptoKit

/// Phone↔burner pairing — the burner (initiator) side.
///
/// The burner shows a QR + a short human code. The phone scans/types it,
/// generates its own X25519 keypair, derives a shared secret, and both
/// sides compute the SAME 6-digit SAS + AES-256-GCM key LOCALLY. The
/// burner forwards its ephemeral public key over the relay (`burner-hello`)
/// so a typed-code phone (which only has the short code, not the pubkey)
/// can complete the handshake; a QR phone already has it from the QR and
/// verifies the relayed copy matches.
///
/// Crypto is byte-compatible with the phone's `QrRelay` (the website QR
/// flow): X25519 → HKDF-SHA256 with salt `flagship/qr/v1`, info
/// `flagship/qr/sas/v1` (SAS) and `flagship/qr/enc/v1` (AEAD key). The
/// only burner-specific addition is the short-code → session-id mapping
/// (`flagship/burner-sid/v1`), pinned by a cross-platform vector.
///
/// SECURITY: the SAS is never on the wire. A relay that swapped pubkeys
/// produces a mismatched code on the two screens, which the user catches.
/// The live socket is the gate — when it drops, the burner re-locks.
public enum BurnerPairing {

    public static let relayHkdfSalt = Data("flagship/qr/v1".utf8)
    public static let encInfo = Data("flagship/qr/enc/v1".utf8)
    public static let sasInfo = Data("flagship/qr/sas/v1".utf8)
    /// Domain-separation tag for deriving the 32-char relay session id from
    /// the short human code. MUST match the phone side.
    public static let sidTag = Data("flagship/burner-sid/v1".utf8)

    /// Bytes of entropy behind the short human code. 5 bytes = exactly 8
    /// base32 chars (no padding) — short enough to read aloud / type, and
    /// (with the relay's per-IP upgrade limit) ample for a short-lived
    /// pairing whose real authentication is the SAS.
    public static let codeByteCount = 5

    public enum PairingError: Error, LocalizedError {
        case badPublicKey
        case derivationFailed(String)
        case badShortCode
        case openFailed(String)
        public var errorDescription: String? {
            switch self {
            case .badPublicKey: return "Phone public key must be 32 raw X25519 bytes."
            case .derivationFailed(let m): return "Shared-secret derivation failed: \(m)"
            case .badShortCode: return "That code isn't valid — check it on the burner."
            case .openFailed(let m): return "Couldn't decrypt what the phone sent: \(m)"
            }
        }
    }

    // MARK: - Session identity (code ⇄ sid)

    /// Generate the random bytes behind a new pairing's short code.
    public static func newCodeBytes() -> Data {
        Data((0..<codeByteCount).map { _ in UInt8.random(in: 0...255) })
    }

    /// The relay session id derived from the short code's bytes — the
    /// `<sid>` in `/burner-pipe/<sid>`. base64url of SHA256(tag || code),
    /// truncated to 32 chars (within the relay's 16–64 bound). Both peers
    /// compute it identically, so a QR (which could also carry the sid) and
    /// a typed code converge on the same Durable Object.
    public static func sessionId(forCodeBytes code: Data) -> String {
        var h = SHA256()
        h.update(data: sidTag)
        h.update(data: code)
        let digest = Data(h.finalize())
        return String(Base64URLBurner.encode(digest).prefix(32))
    }

    /// The human-typeable short code (8 uppercase base32 chars, no padding).
    public static func humanCode(fromBytes code: Data) -> String {
        Base32.encode(code)
    }

    /// Decode a user-typed short code back to its bytes. Tolerant of
    /// lowercase, spaces, and dashes (display grouping). nil if invalid.
    public static func codeBytes(fromHumanCode raw: String) -> Data? {
        let cleaned = raw.uppercased().filter { $0 != " " && $0 != "-" }
        guard let bytes = Base32.decode(cleaned), bytes.count == codeByteCount else { return nil }
        return bytes
    }

    /// Display form of the short code, dash-grouped in 4s: "ABCD-EFGH".
    public static func formatHumanCode(_ code: String) -> String {
        let chars = Array(code)
        guard chars.count == 8 else { return code }
        return String(chars[0..<4]) + "-" + String(chars[4..<8])
    }

    /// The QR payload the burner displays. Carries the short code AND the
    /// burner's ephemeral X25519 pubkey so a scanning phone needs no relay
    /// round-trip to derive the SAS.
    public static func qrPayload(humanCode: String, burnerPublicKey: Data) -> String {
        "flagship://burner?c=\(humanCode)&k=\(Base64URLBurner.encode(burnerPublicKey))"
    }

    // MARK: - Handshake

    public struct DerivedMaterial {
        public let matchCode: String        // 6 digits, zero-padded
        public let aeadKey: SymmetricKey
    }

    /// Burner-side X25519 + HKDF. `burnerPrivateKey` is this session's
    /// ephemeral key; `phonePublicKey` is the raw 32-byte key the phone
    /// sent in `phone-hello`. Mirrors `QrRelay.deriveMaterial` exactly.
    public static func deriveMaterial(
        burnerPrivateKey: Curve25519.KeyAgreement.PrivateKey,
        phonePublicKey: Data
    ) throws -> DerivedMaterial {
        guard phonePublicKey.count == 32 else { throw PairingError.badPublicKey }
        let peer: Curve25519.KeyAgreement.PublicKey
        do { peer = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: phonePublicKey) }
        catch { throw PairingError.badPublicKey }

        let shared: SharedSecret
        do { shared = try burnerPrivateKey.sharedSecretFromKeyAgreement(with: peer) }
        catch { throw PairingError.derivationFailed("\(error)") }

        let aeadKey = shared.hkdfDerivedSymmetricKey(
            using: SHA256.self, salt: relayHkdfSalt, sharedInfo: encInfo, outputByteCount: 32)
        let sasBytes = shared.hkdfDerivedSymmetricKey(
            using: SHA256.self, salt: relayHkdfSalt, sharedInfo: sasInfo, outputByteCount: 4
        ).withUnsafeBytes { Data($0) }

        let u32 = UInt32(sasBytes[0]) << 24 | UInt32(sasBytes[1]) << 16
                | UInt32(sasBytes[2]) << 8 | UInt32(sasBytes[3])
        return DerivedMaterial(matchCode: String(format: "%06d", u32 % 1_000_000), aeadKey: aeadKey)
    }

    /// AEAD-open a payload the phone sealed with the session key. The phone
    /// ships `ciphertext (= ct||tag)` and a 12-byte `nonce` separately
    /// (matching `QrRelay.seal`).
    public static func open(
        ciphertextBase64Url ct: String,
        nonceBase64Url nonce: String,
        key: SymmetricKey
    ) throws -> Data {
        guard let ctData = Base64URLBurner.decode(ct), let nonceData = Base64URLBurner.decode(nonce),
              nonceData.count == 12, ctData.count >= 16 else {
            throw PairingError.openFailed("malformed ciphertext/nonce")
        }
        let tag = ctData.suffix(16)
        let body = ctData.prefix(ctData.count - 16)
        do {
            let box = try AES.GCM.SealedBox(nonce: try AES.GCM.Nonce(data: nonceData),
                                            ciphertext: body, tag: tag)
            return try AES.GCM.open(box, using: key)
        } catch {
            throw PairingError.openFailed("\(error)")
        }
    }

    /// Display split of the 6-digit SAS ("123 456").
    public static func formatMatchCode(_ code: String) -> String {
        guard code.count == 6 else { return code }
        let i = code.index(code.startIndex, offsetBy: 3)
        return code[code.startIndex..<i] + " " + code[i...]
    }
}

/// base64url (RFC 4648 §5, no padding) — local copy so FlagshipBurnerCore
/// has no cross-package dependency.
public enum Base64URLBurner {
    public static func encode(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
    public static func decode(_ s: String) -> Data? {
        var p = s.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        p.append(String(repeating: "=", count: (4 - p.count % 4) % 4))
        return Data(base64Encoded: p)
    }
}

/// RFC 4648 base32 (uppercase A–Z2–7, no padding). Used for the short
/// human pairing code — case-insensitive on decode, dash/space tolerant.
public enum Base32 {
    private static let alphabet = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZ234567")

    public static func encode(_ data: Data) -> String {
        var out = ""
        var buffer = 0
        var bits = 0
        for byte in data {
            buffer = (buffer << 8) | Int(byte)
            bits += 8
            while bits >= 5 {
                bits -= 5
                out.append(alphabet[(buffer >> bits) & 0x1f])
            }
        }
        if bits > 0 {
            out.append(alphabet[(buffer << (5 - bits)) & 0x1f])
        }
        return out
    }

    public static func decode(_ s: String) -> Data? {
        var lookup = [Character: Int]()
        for (i, c) in alphabet.enumerated() { lookup[c] = i }
        var out = Data()
        var buffer = 0
        var bits = 0
        for ch in s {
            guard let v = lookup[ch] else { return nil }
            buffer = (buffer << 5) | v
            bits += 5
            if bits >= 8 {
                bits -= 8
                out.append(UInt8((buffer >> bits) & 0xff))
            }
        }
        return out
    }
}

File: packages/protocol/src/burnerPairing.ts
/**
 * Phone↔burner pairing crypto — the burner (initiator) side, in TS.
 *
 * This is the cross-platform mirror of apps/burner-mac BurnerPairing.swift
 * (Swift) + apps/mobile/android core/BurnerPairing.kt (Kotlin). It lets a
 * Node burner (packages/flagship-burner `pair`) run the SAME pairing the
 * macOS Swift burner does, so a Linux / Chromebook box can pair with the
 * phone and receive the recipe.
 *
 * Crypto is byte-compatible with the QR-relay stack: X25519 → HKDF-SHA256
 * with salt `flagship/qr/v1`, info `flagship/qr/sas/v1` (SAS) and
 * `flagship/qr/enc/v1` (AES-256-GCM key). The short-code → session-id
 * mapping uses `flagship/burner-sid/v1`. All pinned by burnerPairing.test.ts
 * (and apps/com/test/burnerPairingVector.test.ts).
 */
import { x25519 } from "@noble/curves/ed25519.js";
import { hkdf } from "@noble/hashes/hkdf";
import { sha256 } from "@noble/hashes/sha256";
import { gcm } from "@noble/ciphers/aes";
import type { Bytes } from "./types.js";

const enc = (s: string) => new TextEncoder().encode(s);
const RELAY_SALT = enc("flagship/qr/v1");
const ENC_INFO = enc("flagship/qr/enc/v1");
const SAS_INFO = enc("flagship/qr/sas/v1");
const SID_TAG = enc("flagship/burner-sid/v1");
const CODE_BYTE_COUNT = 5;

// ── base64url (no padding) ──────────────────────────────────────────────
const B64_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
export function base64UrlEncode(bytes: Bytes): string {
  let out = "";
  for (let i = 0; i < bytes.length; i += 3) {
    const b0 = bytes[i]!;
    const b1 = i + 1 < bytes.length ? bytes[i + 1]! : 0;
    const b2 = i + 2 < bytes.length ? bytes[i + 2]! : 0;
    out += B64_ALPHABET[b0 >> 2];
    out += B64_ALPHABET[((b0 & 3) << 4) | (b1 >> 4)];
    if (i + 1 < bytes.length) out += B64_ALPHABET[((b1 & 15) << 2) | (b2 >> 6)];
    if (i + 2 < bytes.length) out += B64_ALPHABET[b2 & 63];
  }
  return out;
}
export function base64UrlDecode(s: string): Bytes | null {
  const lookup = new Map<string, number>();
  for (let i = 0; i < B64_ALPHABET.length; i++) lookup.set(B64_ALPHABET[i]!, i);
  const out: number[] = [];
  let buffer = 0;
  let bits = 0;
  for (const ch of s) {
    const v = lookup.get(ch);
    if (v === undefined) return null;
    buffer = (buffer << 6) | v;
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      out.push((buffer >> bits) & 0xff);
    }
  }
  return Uint8Array.from(out);
}

// ── base32 RFC 4648 (uppercase A–Z2–7, no padding) ──────────────────────
const B32_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
export function base32Encode(bytes: Bytes): string {
  let out = "";
  let buffer = 0;
  let bits = 0;
  for (const b of bytes) {
    buffer = (buffer << 8) | b;
    bits += 8;
    while (bits >= 5) {
      bits -= 5;
      out += B32_ALPHABET[(buffer >> bits) & 0x1f];
    }
  }
  if (bits > 0) out += B32_ALPHABET[(buffer << (5 - bits)) & 0x1f];
  return out;
}
export function base32Decode(s: string): Bytes | null {
  const out: number[] = [];
  let buffer = 0;
  let bits = 0;
  for (const ch of s) {
    const v = B32_ALPHABET.indexOf(ch);
    if (v < 0) return null;
    buffer = (buffer << 5) | v;
    bits += 5;
    if (bits >= 8) {
      bits -= 8;
      out.push((buffer >> bits) & 0xff);
    }
  }
  return Uint8Array.from(out);
}

// ── session identity ────────────────────────────────────────────────────
export function newCodeBytes(): Bytes {
  const b = new Uint8Array(CODE_BYTE_COUNT);
  crypto.getRandomValues(b);
  return b;
}

export function sessionId(codeBytes: Bytes): string {
  const input = new Uint8Array(SID_TAG.length + codeBytes.length);
  input.set(SID_TAG, 0);
  input.set(codeBytes, SID_TAG.length);
  return base64UrlEncode(sha256(input)).slice(0, 32);
}

export function humanCode(codeBytes: Bytes): string {
  return base32Encode(codeBytes);
}

export function formatHumanCode(code: string): string {
  return code.length === 8 ? `${code.slice(0, 4)}-${code.slice(4)}` : code;
}

export function codeBytesFromHuman(raw: string): Bytes | null {
  const cleaned = raw.toUpperCase().replace(/[ -]/g, "");
  const bytes = base32Decode(cleaned);
  return bytes && bytes.length === CODE_BYTE_COUNT ? bytes : null;
}

export function qrPayload(human: string, burnerPub: Bytes): string {
  return `flagship://burner?c=${human}&k=${base64UrlEncode(burnerPub)}`;
}

// ── handshake ───────────────────────────────────────────────────────────
export interface BurnerKeypair {
  secretKey: Bytes;
  publicKey: Bytes;
}
export function newBurnerKeypair(): BurnerKeypair {
  const secretKey = x25519.utils.randomSecretKey();
  return { secretKey, publicKey: x25519.getPublicKey(secretKey) };
}

export interface SessionMaterial {
  sasCode: string; // 6 digits, zero-padded
  aeadKey: Bytes; // 32-byte AES-256-GCM key
}
export function deriveSessionMaterial(burnerSecret: Bytes, phonePublicKey: Bytes): SessionMaterial {
  const shared = x25519.getSharedSecret(burnerSecret, phonePublicKey);
  const aeadKey = hkdf(sha256, shared, RELAY_SALT, ENC_INFO, 32);
  const sas4 = hkdf(sha256, shared, RELAY_SALT, SAS_INFO, 4);
  const u32 =
    ((sas4[0]! << 24) | (sas4[1]! << 16) | (sas4[2]! << 8) | sas4[3]!) >>> 0;
  return { sasCode: String(u32 % 1_000_000).padStart(6, "0"), aeadKey };
}

export function formatSas(code: string): string {
  return code.length === 6 ? `${code.slice(0, 3)} ${code.slice(3)}` : code;
}

/** Open a phone-delivered payload: ciphertext (ct||tag) + a 12-byte nonce,
 *  both base64url (matching the phone's seal). Throws on a bad tag. */
export function openDelivered(ciphertextB64u: string, nonceB64u: string, aeadKey: Bytes): Bytes {
  const ct = base64UrlDecode(ciphertextB64u);
  const nonce = base64UrlDecode(nonceB64u);
  if (!ct || !nonce || nonce.length !== 12) throw new Error("malformed delivery");
  return gcm(aeadKey, nonce).decrypt(ct);
}

/** Seal a payload as the phone would (for tests / a phone simulator). */
export function sealDelivered(plaintext: Bytes, aeadKey: Bytes): { ciphertextB64u: string; nonceB64u: string } {
  const nonce = new Uint8Array(12);
  crypto.getRandomValues(nonce);
  const ct = gcm(aeadKey, nonce).encrypt(plaintext);
  return { ciphertextB64u: base64UrlEncode(ct), nonceB64u: base64UrlEncode(nonce) };
}
